src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultFileUriProvider.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.IO;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Shared]
+    [Export(typeof(FileUriProvider))]
+    internal class DefaultFileUriProvider : FileUriProvider
+    {
+        private readonly ITextDocumentFactoryService _textDocumentFactory;
+
+        [ImportingConstructor]
+        public DefaultFileUriProvider(ITextDocumentFactoryService textDocumentFactory)
+        {
+            if (textDocumentFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textDocumentFactory));
+            }
+
+            _textDocumentFactory = textDocumentFactory;
+        }
+
+        public override Uri GetOrCreate(ITextBuffer textBuffer)
+        {
+            if (textBuffer is null)
+            {
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            string filePath;
+            if (_textDocumentFactory.TryGetTextDocument(textBuffer, out var textDocument))
+            {
+                filePath = textDocument.FilePath;
+            }
+            else
+            {
+                // TextBuffer doesn't have a file path, we need to fabricate one.
+                filePath = Uri.UriSchemeFile + Uri.SchemeDelimiter + Guid.NewGuid().ToString();
```

Comment:
>filePath = Uri.UriSchemeFile + Uri.SchemeDelimiter + Guid.NewGuid().ToString(); [](start = 16, length = 79)

Should this get stored in the ITextBuffer.Properties? Otherwise you would get a different path for the same ITextBuffer if you called this multiple times, which seems wrong to me.

Reply:
Dope idea. Ya I should do that.

Reply:
Would it make sense to create a textdocument and just use that instead of buffer properties?

---
In reply to: [381428599](https://github.com/dotnet/aspnetcore-tooling/pull/1602#discussion_r381428599) [](ancestors = 381428599)

Reply:
I could however my initial reasons for not creating a TextDocument was because I didn't want to enable any other parts of the VS system that I wasn't aware of. Basically, if VS didn't already create a TextDocument me creating one might result in issues. The end-goal of this API is to just build something to ensure the TextBuffer is addressable.